File: src/solve/Q4963.java
package solve;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class Q4963 {


    static final int VISIT = 1;
    static final int NONE = 0;
    static final int ISLAND = 1;

    static int[][] visited = new int[51][51];
    static int mX;
    static int mY;
    static int[][] map;


    static ArrayList<TestCase> caseList = new ArrayList<TestCase>();
    final static int[] xPosition = {-1, 0, 1, -1, 1, -1, 0, 1};
    final static int[] yPosition = {1, 1, 1, 0, 0, -1, -1, -1};

    public static void main(String args[]) {

        Scanner sc = new Scanner(System.in);

        //문제입력
        while (true) {
            int tmpX = sc.nextInt();
            int tmpY = sc.nextInt();
            if (tmpX == 0 && tmpY == 0) break;
            int[][] tmpMap = new int[tmpX + 1][tmpY + 1];
            for (int j = 1; j <= tmpY; j++) {
                for (int i = 1; i <= tmpX; i++) {
                    tmpMap[i][j] = sc.nextInt();
                }
            }
            TestCase tmpCase = new TestCase();
            tmpCase.x = tmpX;
            tmpCase.y = tmpY;
            tmpCase.map = tmpMap;
            caseList.add(tmpCase);
        }

        //문제풀이
        for (TestCase testCase : caseList) {
            int count = 0;
            visitedInit();
            mX = testCase.x;
            mY = testCase.y;
            map = testCase.map;
            for (int j = 1; j <= testCase.y; j++) {
                for (int i = 1; i <= testCase.x; i++) {
                    if (visited[i][j] == NONE && map[i][j] == ISLAND) {
                        count++;
                        search(i, j);
                    }
                }
            }
            System.out.println(count);
        }
    }

    static void search(int x, int y) {
        if (visited[x][y] == VISIT) return;
        visited[x][y] = VISIT;
        for (int i = 0; i <= 7; i++) {
            int cX = x + xPosition[i];
            int cY = y + yPosition[i];
            if (isMove(cX, cY) && map[cX][cY] == ISLAND)
                search(cX, cY);
        }
    }

    static boolean isMove(int x, int y) {
        return (x >= 0 && y >= 0 && x <= mX && y <= mY);
    }

    static void visitedInit() {
        for (int i = 0; i <= 50; i++) {
            Arrays.fill(visited[i], 0);
        }

    }

    static class TestCase {
        int x;
        int y;
        int[][] map;
    }


}

File: src/Q4963.kt
import java.util.*
import kotlin.collections.ArrayList

object Q4963 {


    val IMPOSSIBLE = -1
    val POSSIBLE = 1
    val NONE = 0;


    var cache = Array(50) { IntArray(50) }
    var mX = 0;
    var mY = 0;


    @JvmStatic
    fun main(args: Array<String>) {
        val caseList = ArrayList<Case>()
        val sc = Scanner(System.`in`)

        while (true) {
            val tmpCase = Case()
            tmpCase.x = sc.nextInt()
            tmpCase.y = sc.nextInt()
            tmpCase.map = Array(tmpCase.x) { IntArray(tmpCase.y) }
            if (tmpCase.x == 0 && tmpCase.y == 0)
                break
            for (i in 0 until tmpCase.y) {
                for (j in 0 until tmpCase.x) {
                    tmpCase.map!![i][j] = sc.nextInt()
                }
            }
        }

        for (case in caseList) {
            mX = case.x
            mY = case.y
        }


    }


    fun search(x: Int, y: Int) {

        // 움직일 수 있는 8가지 방향!
        // x+1 y
        // x-1 y

        // x y+1
        // x y-1

        // x+1 y+1
        // x+1 y-1

        // x-1 y+1
        // x-1 y-1

        if (cache[x][y] != NONE)
            return
        //시작은 각 경로를 탐색하자

        if (isMove(x + 1, y)) {
            cache[x + 1][y] = 1
            search(x + 1, y)
        } else {
            cache[x + 1][y] = 0;
        }
        if (isMove(x - 1, y)) search(x - 1, y)
        if (isMove(x, y + 1)) search(x, y + 1)
        if (isMove(x, y - 1)) search(x, y - 1)
        if (isMove(x + 1, y + 1)) search(x + 1, y + 1)
        if (isMove(x + 1, y - 1)) search(x + 1, y - 1)
        if (isMove(x - 1, y + 1)) search(x - 1, y + 1)
        if (isMove(x - 1, y - 1)) search(x - 1, y - 1)


    }


    fun isMove(x: Int, y: Int): Boolean {
        return (x <= mX && y <= mY)
    }

}


data class Case(
        var map: Array<IntArray>? = null,
        var x: Int = 0,
        var y: Int = 0

) {}
